segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/hdfs/HDFSStorage.java

```diff
@@ -196,85 +323,309 @@ public boolean supportsTruncation() {
+                // Note: IOUtils.copyBytes with length == 0 will enter an infinite loop, hence the need for this check.
+                return;
+            }
+
+            // We need to be very careful with IOUtils.copyBytes. There are many overloads with very similar signatures.
+            // There is a difference between (InputStream, OutputStream, int, boolean) and (InputStream, OutputStream, long, boolean),
+            // in that the one with "int" uses the third arg as a buffer size, and the one with "long" uses it as the number
+            // of bytes to copy.
+            IOUtils.copyBytes(data, stream, (long) length, false);
+
+            stream.flush();
+        } catch (IOException ex) {
+            throw HDFSExceptionHelpers.convertException(handle.getSegmentName(), ex);
+        }
+
+        HDFSMetrics.WRITE_LATENCY.reportSuccessEvent(timer.getElapsed());
+        HDFSMetrics.WRITE_BYTES.add(length);
+        LoggerHelpers.traceLeave(log, "write", traceId, handle, offset, length);
     }
 
     @Override
-    public SegmentProperties getStreamSegmentInfo(String streamSegmentName) throws StreamSegmentException {
-        return call(new GetInfoOperation(streamSegmentName, this.context));
+    public SegmentHandle openWrite(String streamSegmentName) throws StreamSegmentException {
+        long traceId = LoggerHelpers.traceEnter(log, "openWrite", streamSegmentName);
+        int fencedCount = 0;
+        do {
+            try {
+                FileStatus fileStatus = findStatusForSegment(streamSegmentName, true);
+
+                if (!isSealed(fileStatus.getPath())) {
+                    if (getEpochFromPath(fileStatus.getPath()) > this.epoch) {
+                        throw new StorageNotPrimaryException(streamSegmentName);
+                    }
+
+                    Path targetPath = getFilePath(streamSegmentName, this.epoch);
+                    if (!targetPath.equals(fileStatus.getPath())) {
+                        try {
+                            this.fileSystem.rename(fileStatus.getPath(), targetPath);
+                        } catch (FileNotFoundException e) {
+                            //This happens when more than one host is trying to fence and only one of the host goes through.
+                            //Retry the rename so that host with the highest epoch gets access.
+                            log.warn("Race in fencing. More than two hosts trying to own the segment. Retrying");
+                            fencedCount++;
+                            continue;
+                        }
+                    }
+                }
+                //Ensure that file exists
+                findStatusForSegment(streamSegmentName, true);
+                return HDFSSegmentHandle.write(streamSegmentName);
+            } catch (IOException e) {
+                throw HDFSExceptionHelpers.convertException(streamSegmentName, e);
+            }
+        } while (fencedCount < MAX_ATTEMPT_COUNT);
```

Comment:
This will happen when all the hosts are trying to get the same file and the authoritative owner always ends up losing the race at each iteration. Both these events are very rare.
I'd like to fix this in `HDFSStorage` itself. Throwing this issue to the caller will complicate the caller design. Here is one proposed fix: If we set the var `MAX_ATTEMPT_COUNT` to a value far larger than the number of hosts then we are sure that the real owner never loses. My recommendation is to put this value to 500. How does this sound?

Reply:
> Both these events are very rare.

I agree, but major outages are often caused out of rare events.

> I'd like to fix this in HDFSStorage itself. Throwing this issue to the caller will complicate the caller design.

I agree.

> Here is one proposed fix: If we set the var MAX_ATTEMPT_COUNT to a value far larger than the number of hosts then we are sure that the real owner never loses. My recommendation is to put this value to 500. How does this sound?

I think this is making the implicit assumption that older owners will start first and consequently will exhaust their attempts first, but that's not necessarily true. A segment container owner might be elected and then say pause for arbitrarily long. A new owner can start first.

The property we need to preserve is the following:
- If an instance is the last designated owner of a segment container, then it needs to eventually succeed in the call to rename to a higher epoch.

Basically, if an instance has retried enough and still couldn't make it, then it shouldn't give up in the case it is the authoritative owner according to the controller. What needs to happen is that the instance needs to be able to go back and check that it is still the designated owner to determine whether it should keep retrying or give up. If I recall correctly, we don't have a way to check it from the segment store at the moment, which makes this difficult.

Reply:
If the new segmentstores (the owners with higher epochs) win the race, the older segmentstores leave the race. So in worst case, at least one segmentstore is out of the race. So at the max the correct owner needs to loop for number of hosts times before he wins the race. I think this loop is bounded and will succeed eventually.
In case of pausing segmentstores, as rename is atomic, change in ownership will fail any further operations on the segment.
Consider an example with 5 hosts. I think each host running the loop 5 times will ensure that the latest owner wins the race always.

Reply:
@arvindkandhare did you mean to elaborate on an example?

> Consider an example with 5 hosts.

It might actually be sufficient to walk through an example with 2 or 3 hosts. I'm still not sure how you guarantee this:

> So at the max the correct owner needs to loop for number of hosts times

Reply:
Consider 3 hosts A,B,C with epochs 1,2 and 3. The segment name is "foo".The file is currently owned by epoch 0. (i.e. current file name is "foo_0").
Now in case of a race, all three would try to rename the file to "foo_1", "foo_2" and "foo_3". If rename to "foo_1" goes through, segment store 1 leaves the race as it gets the ownership.
So in case of race during rename, at least one segment store leaves the race as it is the owner. So in worst case the real onwer needs to loop for number of segmentstores in the race to win it eventually. Which can be generalized to total number of segmentstores.

Reply:
Ok, got it. It sounds right that the number of attempts should be max number of segment store instances. We might need to optimize it when clusters grow large, but as a first approximation, it seems ok.

I think we re agreeing that the best approach would to set the maximum based on the size of the cluster rather than hardcoding a value? Again, if you want to do it separately, fine by me, but I'd like to have an issue to fix it once we agree what the right fix is.

Reply:
Here are the issues to track the issues coming out of this thread and offline discussions: #2501, #2503.

Reply:
I saw there was an issue (#2501) created for this. I disagree with this approach. The Tier2 adapter does not need to know any details above its layer, especially something that high such as "number of segment store instances. The segment store itself does not know that number either and couldn't care less about it. 

Trying to inject such a value into the Tier2 adapters would be really hard, and it would be even harder trying to keep it accurate. We can bring up and stop segment store instances at will.

I suggest you find a different solution to this problem.

Reply:
@andreipaduroiu please discuss it in the issue, not here.